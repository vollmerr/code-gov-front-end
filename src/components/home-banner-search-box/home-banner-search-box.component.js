import React, { Component } from 'react'
import { map, some } from '@code.gov/cautious'
import Autocomplete from 'components/autocomplete'
import SearchBox from 'components/search-box'
import client from 'api-client'

export default class HomeBannerSearchBoxComponent extends Component {

  constructor(props) {
    super(props)
    this.state = {
      showAutocomplete: false,
      suggestions: []
    }
  }

  componentDidMount() {
    this.mounted = true
  }

  componentWillUnmount() {
    this.mounted = false
  }

  handleBlur = () => {
    if (this.mounted) {
      this.setState({ showAutocomplete: false })
    }
  }

  handleFocus = () => {
    if (this.mounted) {
      this.setState({ showAutocomplete: true })
    }
  }

  handleChange = (value) => {
    client.suggest(value, 5).then(terms => {
      const suggestions = map(terms, term => {
        return {
          text: term,
          to: `/search?query=${term}&page=1&size=1000`
        }
      })
      if (this.mounted) {
        this.setState({
          showAutocomplete: true,
          suggestions
        })
      }
    })
  }

  handleClick = ({text, to}) => {
    this.props.onSubmit(text)
  }

  render() {
    const {searchDescriptionText, searchDescriptionTextMobile, onSubmit, placeholder, query} = this.props
    return (
      <div className="search-input-container">
        <div className="search-input-wrapper">
          <div className="search-description-wrapper">
            {searchDescriptionText && <div className="show-w-gt-800">{searchDescriptionText}</div>}
            {searchDescriptionTextMobile && <div className="show-w-lte-800">{searchDescriptionTextMobile}</div>}
          </div>
          <div className="search-input-and-button-wrapper">
            <SearchBox
              placeholder={placeholder}
              onBlur={this.handleBlur}
              onChange={this.handleChange}
              onFocus={this.handleFocus}
              onSubmit={onSubmit}
              query={query}
            />
            {this.state.showAutocomplete && some(this.state.suggestions) && <Autocomplete options={this.state.suggestions} onClick={this.handleClick}/>}
          </div>
        </div>
      </div>
    )
  }
}
