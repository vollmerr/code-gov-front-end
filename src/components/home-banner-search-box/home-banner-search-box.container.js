import { connect } from 'react-redux'
import { push } from 'connected-react-router'
import updateSearchParams from 'actions/update-search-params'
import HomeBannerSearchBoxComponent from './home-banner-search-box.component'
import { getConfigValue } from 'utils/other'
import { getSection } from 'utils/url-parsing'


export const mapStateToProps = ({ query }) => {
  return {
    placeholder: getConfigValue('content.home.banner.search_placeholder_text'),
    searchDescriptionText: getConfigValue('content.home.banner.search_description_text'),
    searchDescriptionTextMobile: getConfigValue('content.home.banner.search_description_text_mobile'),
    query
  }
}

export const mapDispatchToProps = dispatch => {
  return {
    onSubmit: (query) => {
      console.log("home-banner-search-box.container starting onSubmit with query:", query)
      if (getSection === 'search') {
        dispatch(updateSearchParams({ page: 1, query, size: 10 }))
      } else {
        dispatch(updateSearchParams({ page: 1, query, size: 10, sort: 'best_match' }))
        dispatch(push(`/search?page=1&query=${query}&size=1000&sort=best_match`))
      }
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(HomeBannerSearchBoxComponent)
