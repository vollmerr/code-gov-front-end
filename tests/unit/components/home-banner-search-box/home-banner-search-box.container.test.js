import { push } from 'connected-react-router'

import updateSearchParams from 'actions/update-search-params'
import { mapStateToProps, mapDispatchToProps } from 'components/home-banner-search-box/home-banner-search-box.container'

jest.mock('connected-react-router')
jest.mock('actions/update-search-params')

const props = {
  query: 'test-query',
}

const dispatch = jest.fn()

describe('containers - HomeBannerSearchBox', () => {
  describe('mapStateToProps', () => {
    it('should return the correct properties', () => {
      expect(mapStateToProps(props)).toMatchSnapshot()
    })
  })

  describe('mapDispatchToProps', () => {
    describe('onSubmit', () => {
      describe('section is `search`', () => {
        xit('should dispatch the `updateSearchParams` action with the correct params', () => {
          // BUG: missing call to getSection...?
        })
      })

      describe('section is not `search`', () => {
        it('should dispatch the `updateSearchParams` action with the correct params', () => {
          mapDispatchToProps(dispatch).onSubmit(props.query)
          expect(dispatch).toBeCalled()
          expect(updateSearchParams).toBeCalledWith({ page: 1, query: props.query, size: 1000, sort: 'best_match' })
        })

        it('should dispatch the `push` action with the correct params', () => {
          mapDispatchToProps(dispatch).onSubmit(props.query)
          expect(dispatch).toBeCalled()
          expect(push).toBeCalledWith(`/search?page=1&query=${props.query}&size=1000&sort=best_match`)
        })
      })
    })
  })
})
