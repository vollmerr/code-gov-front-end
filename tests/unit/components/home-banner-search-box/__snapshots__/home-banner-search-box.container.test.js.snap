// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers - HomeBannerSearchBox mapStateToProps should return the correct properties 1`] = `
Object {
  "placeholder": "e.g. JavaScript, NASA, web standards",
  "query": "test-query",
  "searchDescriptionText": "Search thousands of California State projects",
  "searchDescriptionTextMobile": "Search projects",
}
`;
